src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -85,12 +84,36 @@ public TypeFacetsPromotionRules TypeFacetsPromotionRules
         /// <returns>The resolved navigation source.</returns>
         public virtual IEdmNavigationSource ResolveNavigationSource(IEdmModel model, string identifier)
         {
+            ExceptionUtils.CheckArgumentNotNull(model, nameof(model));
+            ExceptionUtils.CheckArgumentNotNull(identifier, nameof(identifier));
+
             IEdmNavigationSource navSource = model.FindDeclaredNavigationSource(identifier);
             if (navSource != null || !EnableCaseInsensitive)
             {
                 return navSource;
             }
 
+            if (model.IsImmutable())
+            {
+                NormalizedModelElementsCache cache = GetNormalizedModelElementsCache(model);
+                IList<IEdmNavigationSource> cachedResults = cache.FindNavigationSources(identifier);
+
+                if (cachedResults != null)

```

Comment:
This is the reason why it's preferrable to return an empty collection rather than null from a function that has a collection return type.

Reply:
While I agree that it's generally preferrable, return an empty collection for the cache would introduce more complexity to the code than this single null-check that's not exposed to the user.

I would have to create a static empty list for each element type that the cache supports:
```c#
static readonly List<IEdmSchemaType> emptySchemaTypesList = new List<IEdmSchemaType>();
static readonly List<IEdmOperation> emptyOperationsList = new List<...>;
static readonly List<IEdmTerm> emptyTermsList = ...;
static readonly List<IEdmNavigationSource> emptyNavigationSourcesList = ...;
static readonly List<IEdmOperationImport> emptyOperationImportsList = ...;

// then in the find methods:
if (navigationSources.TryGetValue(...))
{
      return results;
}

return emptyNavigationSourcesList;
```

The runtime overhead is not that bad since these collections are allocated only once, but it didn't feel worthwhile since the cache is an internal implementation detail used in a very limited scope.

That said, I don't hold this opinion strongly. If you still believe that I should return an empty collection despite the explanation above, I can go ahead and make the change.